### PR TITLE
[PWGJE] First update heavy flavour tagging analysis

### PR DESCRIPTION
**Pull request by hanseopark**
Core/JetTaggingUtilities.h
- Add calculation of geometric sign which is product dot jet momentum and dca momentum
- 
DataModel/JetTagging.h
- Add propagation of track for dca_X and dcaY
- Add table of tracks which have parameter for track counting. it is expected to expend different method

TableProducer/jettaggerhf.cxx
- Making table for tagging jet and tracks

TableProducer/jettaggerhfExtension.cxx
- Add propagation of track for dca_X and dca_Y

Tasks/jettaggerhfTC.cxx
- First QA for hf jet tagging for flavour decision and comparing track's parameters

### CODE REVIEW

File: PWGJE/Core/JetTaggingUtilities.h [hunk 1]
```diff
@@ -38,6 +39,7 @@ enum JetTaggingSpecies {
 
 namespace JetTaggingUtilities
 {
+const int cmTomum = 10000;
```
Comment (nzardosh) [inline PWGJE/Core/JetTaggingUtilities.h:42]: can you put a comment here of what this is please?

File: PWGJE/Core/JetTaggingUtilities.h [hunk 2]
```diff
@@ -275,6 +278,22 @@ int jetOrigin(T const& jet, U const& particles, float dRMax = 0.25)
...
+ *
+ * @param collision which is needed external table of collision due to postion X and Y
+ * @param jet
+ * @param track which is needed each DCA_X and Y which is measured in jettaggerhfExtension.cxx
+ */
+template <typename T, typename U, typename V>
+int getGeoSign(T const& collision, U const& jet, V const& track)
+{
+  auto sign = TMath::Sign(1, (track.dcaX() - collision.posX()) * jet.px() + (track.dcaY() - collision.posY()) * jet.py() + (track.dcaZ() - collision.posZ()) * jet.pz());
+  if (sign < -1 || sign > 1)
```
Comment (nzardosh) [inline PWGJE/Core/JetTaggingUtilities.h:292]: what is this if for?

File: PWGJE/DataModel/JetTagging.h [hunk 3]
```diff
@@ -26,28 +27,143 @@ using namespace o2::analysis;
...
+    DECLARE_SOA_COLUMN(Algorithm2, algorithm2, int);             \
+    DECLARE_SOA_COLUMN(Algorithm3, algorithm3, int);             \
+  }                                                              \
+  DECLARE_SOA_TABLE(_jet_type_##Tags, "AOD", _description_ "TJ", \
+                    _name_##tagging::Origin,                     \
+                    _name_##tagging::Algorithm1,                 \
+                    _name_##tagging::Algorithm2,                 \
+                    _name_##tagging::Algorithm3);
+
+#define DECLARE_TAGTRACK_TABLE(_jet_type_, _name_, _track_type_, _description_)  \
```
Comment (nzardosh) [inline PWGJE/DataModel/JetTagging.h:115]: please use TRACKTAG_TABLE_DEF as the name for consistency

Comment (nzardosh) [reply PWGJE/DataModel/JetTagging.h:115]: same as instances below of other defines

File: PWGJE/TableProducer/jettaggerhf.cxx [hunk 4]
```diff
@@ -21,64 +22,113 @@
...
+#include "PWGJE/Core/FastJetUtilities.h"
 #include "PWGJE/Core/JetTaggingUtilities.h"
 #include "PWGJE/Core/JetDerivedDataUtilities.h"
 
 using namespace o2;
 using namespace o2::framework;
 using namespace o2::framework::expressions;
 
-template <typename JetTableData, typename JetTableMCD, typename JetTaggingTableData, typename JetTaggingTableMCD>
+template <typename JetTableData, typename JetTableMCD, typename JetTaggingTableData, typename ConstTaggingTableData, typename ConstTCTaggingTableData, typename JetTaggingTableMCD, typename ConstTaggingTableMCD, typename ConstTCTaggingTableMCD>
```
Comment (nzardosh) [inline PWGJE/TableProducer/jettaggerhf.cxx:33]: why do the new ones have const in the name?

File: PWGJE/TableProducer/jettaggerhf.cxx [hunk 5]
```diff
@@ -21,64 +22,113 @@
...
+  using JetTagTracksData = soa::Join<aod::JTracks, aod::JTrackTagDcas, aod::JTrackTagDcaCovs>;
+  using JetTagTracksMC = soa::Join<aod::JTracks, aod::JMcTrackLbs, aod::JTrackTagDcas, aod::JTrackTagDcaCovs>;
+
   void processDummy(aod::Collision const& collision)
   {
   }
   PROCESS_SWITCH(JetTaggerHFTask, processDummy, "Dummy process", true);
 
-  void processData(aod::JCollision const& collision, JetTableData const& jets, aod::JTracks const& tracks)
+  void processData(soa::Join<aod::JCollisions, aod::JCollisionPIs>::iterator const& jcollision, aod::Collisions&, JetTableData const& jets, aod::JTracks&, JetTagTracksData const& tracks)
```
Comment (nzardosh) [inline PWGJE/TableProducer/jettaggerhf.cxx:57]: please here use the predefined "using: keyowrds found at the bottom of DataModel/Jet.h

File: PWGJE/TableProducer/jettaggerhf.cxx [hunk 6]
```diff
@@ -21,64 +22,113 @@
...
+        vecTrackPt.push_back(track.pt());
+        vecTrackEta.push_back(track.eta());
+        vecTrackPhi.push_back(track.phi());
+        vecSignedIP2D.push_back(track.dcaXY() / TMath::Sqrt(track.sigmaDcaXY2()) * JetTaggingUtilities::cmTomum);
+        vecSignedIP2Ds.push_back(geoSign * TMath::Abs(track.dcaXY()) / TMath::Sqrt(track.sigmaDcaXY2()) * JetTaggingUtilities::cmTomum);
+        vecSignedIP3D.push_back(track.dcaXYZ() / TMath::Sqrt(track.sigmaDcaXYZ2()) * JetTaggingUtilities::cmTomum);
+        vecSignedIP3Ds.push_back(geoSign * TMath::Abs(track.dcaXYZ()) / TMath::Sqrt(track.sigmaDcaXYZ2()) * JetTaggingUtilities::cmTomum);
+      }
+      constTableData(taggingTableData.lastIndex(), trackconst);
+      constTCTableData(taggingTableData.lastIndex(), 0, jet.pt(), jet.eta(), jet.phi(), vecGeoSign, vecTrackPt, vecTrackEta, vecTrackPhi, vecSignedIP2D, vecSignedIP2Ds, vecSignedIP3D, vecSignedIP3Ds);
```
Comment (nzardosh) [inline PWGJE/TableProducer/jettaggerhf.cxx:85]: why do you save the jet information for each track here? I think its better to remove this since it looks like this table is anyways joinable to the jet table right?

File: PWGJE/TableProducer/jettaggerhf.cxx [hunk 7]
```diff
@@ -21,64 +22,113 @@
...
       // if (doAlgorithm2) algorithm2 = JetTaggingUtilities::Algorithm2((mcdjet, tracks);
       // if (doAlgorithm3) algorithm3 = JetTaggingUtilities::Algorithm3((mcdjet, tracks);
       taggingTableData(0, algorithm1, algorithm2, algorithm3);
+      std::vector<int> trackconst;
+      std::vector<int> vecGeoSign;
+      std::vector<float> vecTrackPt, vecTrackEta, vecTrackPhi, vecSignedIP2D, vecSignedIP2Ds, vecSignedIP3D, vecSignedIP3Ds;
+      for (auto& track : jet.template tracks_as<JetTagTracksData>()) {
+        trackconst.push_back(track.globalIndex());
+        int geoSign = JetTaggingUtilities::getGeoSign(collision, jet, track);
+        trackconst.push_back(track.globalIndex());
```
Comment (nzardosh) [inline PWGJE/TableProducer/jettaggerhf.cxx:74]: why is the track globalIndex pushed back twice?

File: PWGJE/TableProducer/jettaggerhf.cxx [hunk 8]
```diff
@@ -21,64 +22,113 @@
...
+        vecGeoSign.push_back(geoSign);
+        vecTrackPt.push_back(track.pt());
+        vecTrackEta.push_back(track.eta());
```
Comment (nzardosh) [inline PWGJE/TableProducer/jettaggerhf.cxx:77]: this is fine if you intend to store these tables later for offline processing. If not you can just put an index to the track table as you have above aand then you dont need to save pt, eta and phi of tracks

File: PWGJE/DataModel/JetTagging.h [hunk 9]
```diff
@@ -26,28 +27,143 @@ using namespace o2::analysis;
...
+DECLARE_SOA_COLUMN(GeoSign, geoSign, int);
+} // namespace jtracktagsub
+
+namespace constituenttc
+{
+DECLARE_SOA_COLUMN(JetFlavour, jetFlavour, float);
+DECLARE_SOA_COLUMN(JetPt, jetPt, float);
+DECLARE_SOA_COLUMN(JetEta, jetEta, float);
+DECLARE_SOA_COLUMN(JetPhi, jetPhi, float);
+DECLARE_SOA_COLUMN(VecGeoSign, vecGeoSign, std::vector<int>);
```
Comment (nzardosh) [inline PWGJE/DataModel/JetTagging.h:90]: you can check if there is already a predefined array coloumn type or not

File: PWGJE/TableProducer/jettaggerhf.cxx [hunk 10]
```diff
@@ -21,64 +22,113 @@
...
+        vecTrackPhi.push_back(track.phi());
+        vecSignedIP2D.push_back(track.dcaXY() / TMath::Sqrt(track.sigmaDcaXY2()));
+        vecSignedIP2Ds.push_back(geoSign * TMath::Abs(track.dcaXY()) / TMath::Sqrt(track.sigmaDcaXY2()));
+        vecSignedIP3D.push_back(track.dcaXYZ() / TMath::Sqrt(track.sigmaDcaXYZ2()));
+        vecSignedIP3Ds.push_back(geoSign * TMath::Abs(track.dcaXYZ()) / TMath::Sqrt(track.sigmaDcaXYZ2()));
+      }
+      constTableMCD(taggingTableMCD.lastIndex(), trackconst);
```
Comment (nzardosh) [inline PWGJE/TableProducer/jettaggerhf.cxx:122]: since these tables can be joined to the taggingTableMCD I dont think you need to the indices. you can just  join them later

File: PWGJE/TableProducer/jettaggerhfExtension.cxx [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+    }
+    grpmag = ccdb->getForTimeStamp<o2::parameters::GRPMagField>(grpmagPath, bc.timestamp());
+    LOG(info) << "Setting magnetic field to current " << grpmag->getL3Current() << " A for run " << bc.runNumber() << " from its GRPMagField CCDB object";
+    o2::base::Propagator::initFieldFromGRP(grpmag);
+    o2::base::Propagator::Instance()->setMatLUT(lut);
+    mVtx = ccdb->getForTimeStamp<o2::dataformats::MeanVertexObject>(mVtxPath, bc.timestamp());
+    runNumber = bc.runNumber();
+  }
+
+  void calculateDcaXYZ(float& dcaXYZ, float& sigmaDcaXYZ2, float dcaXY, float dcaZ, float cYY, float cZY, float cZZ, float sigmaDcaXY2, float sigmaDcaZ2)
```
Comment (nzardosh) [inline PWGJE/TableProducer/jettaggerhfExtension.cxx:116]: does this not exist already in recodecay?

File: PWGJE/TableProducer/jettaggerhfExtension.cxx [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+// Task to produce a table joinable to the jet tables for hf jet tagging
```
Comment (nzardosh) [inline PWGJE/TableProducer/jettaggerhfExtension.cxx:12]: please fix this

File: PWGJE/TableProducer/jettaggerhfExtension.cxx [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+  {
+    if (runNumber == bc.runNumber()) {
+      return;
+    }
+    grpmag = ccdb->getForTimeStamp<o2::parameters::GRPMagField>(grpmagPath, bc.timestamp());
+    LOG(info) << "Setting magnetic field to current " << grpmag->getL3Current() << " A for run " << bc.runNumber() << " from its GRPMagField CCDB object";
+    o2::base::Propagator::initFieldFromGRP(grpmag);
+    o2::base::Propagator::Instance()->setMatLUT(lut);
+    mVtx = ccdb->getForTimeStamp<o2::dataformats::MeanVertexObject>(mVtxPath, bc.timestamp());
+    runNumber = bc.runNumber();
```
Comment (nzardosh) [inline PWGJE/TableProducer/jettaggerhfExtension.cxx:113]: why is this needed?

File: PWGJE/TableProducer/jettaggerhfExtension.cxx [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+  void initCCDB(T const& bc)
+  {
+    if (runNumber == bc.runNumber()) {
+      return;
+    }
+    grpmag = ccdb->getForTimeStamp<o2::parameters::GRPMagField>(grpmagPath, bc.timestamp());
+    LOG(info) << "Setting magnetic field to current " << grpmag->getL3Current() << " A for run " << bc.runNumber() << " from its GRPMagField CCDB object";
+    o2::base::Propagator::initFieldFromGRP(grpmag);
+    o2::base::Propagator::Instance()->setMatLUT(lut);
+    mVtx = ccdb->getForTimeStamp<o2::dataformats::MeanVertexObject>(mVtxPath, bc.timestamp());
```
Comment (nzardosh) [inline PWGJE/TableProducer/jettaggerhfExtension.cxx:112]: it seems in this function you only need this line?

File: PWGJE/TableProducer/jettaggerhfExtension.cxx [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+          auto const& collision = track.collision();
+          o2::base::Propagator::Instance()->propagateToDCABxByBz({collision.posX(), collision.posY(), collision.posZ()}, trackPar, 2.f, matCorr, &dcaInfo);
+        } else {
+          o2::base::Propagator::Instance()->propagateToDCABxByBz({mVtx->getX(), mVtx->getY(), mVtx->getZ()}, trackPar, 2.f, matCorr, &dcaInfo);
+        }
+      }
+      auto xyz = trackPar.getXYZGlo();
+      float dcaX = 999.;
+      float dcaY = 999.;
+      if (track.has_collision()) {
```
Comment (nzardosh) [inline PWGJE/TableProducer/jettaggerhfExtension.cxx:150]: isnt this already checked in line 131?

File: PWGJE/TableProducer/jettaggerhfExtension.cxx [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+    float dFdz = 2 * dcaZ / dcaXYZ;
+    sigmaDcaXYZ2 = std::abs(cYY * dFdxy * dFdxy + cZZ * dFdz * dFdz + 2 * cZY * dFdxy * dFdz);
+  }
+
+  void processDummy(aod::Collision const& collision)
+  {
+  }
+  PROCESS_SWITCH(JetTaggerHFTrackExtension, processDummy, "Dummy process", true);
+
+  void processTracks(soa::Join<aod::Tracks, aod::TracksCov, aod::TrackSelection, aod::TracksDCA, aod::TracksDCACov, aod::McTrackLabels>::iterator const& track, aod::BCsWithTimestamps&, aod::Collisions&)
```
Comment (nzardosh) [inline PWGJE/TableProducer/jettaggerhfExtension.cxx:129]: what exactly is this task doing? It seems to be recalculating the DCA of tracks but it is doing so with respect to the original collision? dont these coloumns already exist?

**Review by nzardosh — COMMENTED** [review]
=

File: PWGJE/Core/JetTaggingUtilities.h (hunk 1 above)
Comment (hanseopark) [reply PWGJE/Core/JetTaggingUtilities.h:42]: const int cmTomum = 10000; // using cm -> #mum for impact parameter (dca)

I add a comment

File: PWGJE/Core/JetTaggingUtilities.h (hunk 2 above)
Comment (hanseopark) [reply PWGJE/Core/JetTaggingUtilities.h:292]: /**
 * return geometric sign which is calculated scalar product between jet axis with DCA (track propagated to PV )
    * positive and negative value are expected from primary vertex
    * positive value is expected from secondary vertex
    *
    * @param collision which is needed external table of collision due to postion X and Y
    * @param jet
    * @param track which is needed each DCA_X and Y which is measured in jettaggerhfExtension.cxx
    */

File: PWGJE/TableProducer/jettaggerhfExtension.cxx (hunk 14 above)
Comment (hanseopark) [reply PWGJE/TableProducer/jettaggerhfExtension.cxx:112]: I just copy from Common/TableProducer/trackPropagation.cxx
it doesn't need to set magnetic field from ccdb?

File: PWGJE/TableProducer/jettaggerhfExtension.cxx (hunk 15 above)
Comment (hanseopark) [reply PWGJE/TableProducer/jettaggerhfExtension.cxx:150]: is it right. I will change

**Review by hanseopark — COMMENTED** [review]
=

File: PWGJE/DataModel/JetTagging.h (hunk 9 above)
Comment (nzardosh) [reply PWGJE/DataModel/JetTagging.h:90]: No not an index coloumn, but just an array one. But possible there isnt

**Comment by hanseopark** [comment]
I have changed many things which are remove constituents in jet tags, propagation part, add comment and use pre-defined table definition.

DataModel/JetTagging.h
I add extendion of DCA for JTracks becaues I want to use extend from JTracks directly not using originaltrack. I'm not sure why we shouldn't when we can extend JTracks by simply calling the track. Is it much of a memory disadvantage?
If this is possible, I would also like to extend JCollision to use posX and posY as well (stored in the tag data model).

TableProducer/jettaggerhf.cxx
There are saved with flavour information of jet and table for ML.

TableProducer/jettaggerhfExtension.cxx
I have changed that I will not use propagation seperatively because when we use getXYZGlo(), it can be called for propagated track.
So, I did extend JTracks in here.

Task/jettaggerhfTC.cxx
There are task of general QA for hf tagger and analysis of track counting method for prototype.

Let me know if you have any comments

File: PWGJE/TableProducer/jettaggerhf.cxx [hunk 17]
```diff
@@ -12,73 +12,129 @@
...
 using namespace o2;
 using namespace o2::framework;
 using namespace o2::framework::expressions;
 
-template <typename JetTableData, typename JetTableMCD, typename JetTaggingTableData, typename JetTaggingTableMCD>
+template <typename JetTableData, typename JetTableMCD, typename JetTaggingTableData, typename StoredJetTaggingTableData, typename JetTaggingTableMCD, typename StoredJetTaggingTableMCD>
 struct JetTaggerHFTask {
 
   Produces<JetTaggingTableData> taggingTableData;
+  Produces<StoredJetTaggingTableData> storedTaggingTableData;
```
Comment (nzardosh) [inline PWGJE/TableProducer/jettaggerhf.cxx:38]: I think you can change the name of this to something else. Since both this and the one above it are stored. Also the logic seems a bit strange. You have the algorithm outputs in jettaggingtabledata and then you have some variable outputs in storedjettaggingtabledata. However it seems these should be two different tasks and that you should be able to run the trained ml model in this task
